### PR TITLE
Either

### PR DESCRIPTION
**Pull request by boris-marinov**
Most of this is adapted from the Maybe or ported from the old Folktale.

The only part that is authored is the JSVerify tests (seem like an overkill technically but I really like the looks) and a slightly different implementations of `fold` and `swap` (using `cata`). 

This is just a draft - all feedback is welcome.  I myself have some feedback but I am not sure how to structure and share it.

Cheers.

### CODE REVIEW

**Comment by boris-marinov** [comment]
I now realize that I have forgotten to run the linter. Sorry for that. Will fix tomorrow.

File: src/data/either/core.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+// -- Constructors -----------------------------------------------------
+const Left  = (value) => new Either.Left.constructor({ value });
+const Right = (value) => new Either.Right.constructor({ value });
+
+Either.fromNullable = function(a) {
+  return a != null
+    ? Right(a)
+    : Left(a);
+}
```
Comment (robotlolita) [inline src/data/either/core.js]: This should be formatted as:

```js
return a != null ?  Right(a)
:      /* else */   Left(a)
```

See `src/core/lambda/curry.js`

**Comment by robotlolita** [comment]
Sorry, I forgot to push the fixes to Makefile. Could you rebase it on master?

We should move `assert*` to separate modules, so they can be reused. This should probably go into `src/helpers/assert*.js`, but since it would also affect Maybe, we can do that after merging.

`fromNullable` and `try` should be separate modules. This way we can add new utility functions for converting between common structures/etc without burdening people who don't need those functions (this is where I really wish JS had a good support for tree-shaking):

```js
// fromNullable.js
module.exports = (value) => ...

// try.js
module.exports = (computation) => { ... }

// core.js
module.exports = { of: ..., type: ..., ... }

// index.js
module.exports = {
  ...require('./core'),
  fromNullable: require('./fromNullable'),
  try: require('./try')
}
```

Other than that, it looks good to me.

Feel free to share your thoughts. I'm usually not sure how to structure things either, but PRs are meant for discussions, right? :)

**Comment by boris-marinov** [comment]
Having these functions as part of the index would be superb! 

We may want to change the name of  `fromNullable` though, in order for it to make sense in the top-level contex  (maybe we should call it `nullableToEither`). The name of `try` does makes sense, provided that we treat `Either` as a first class citizen (and I think we definitely should).

Also we have the question of how to organize the modules. Maybe we should make them a part of a bigger module called something like `eitherUtils`, which then can be imported on top-level like the `core`: module in your example: 

````javascript
module.exports = {
  ...require('./core'),
  ...require('./eitherUtils')
}
```

I will post more comments later.

**Comment by boris-marinov** [comment]
Seems like the build is OK too.

Waiting for your response on how to organize the `fromNullable` and `try` modules.

**Comment by robotlolita** [comment]
Oh, ok. So since people will be doing requires such as `require('folktale/data/either')` and `require('folktale/data/either/fromNullable')`, I don't think we need to give those functions different names.

As for the functions being part of a bigger module, the reason I'm avoiding that is so people can pick only the functions that they use, directly, when they care about the amount of code they're pulling into their application. This is not a concern for Node.js applications, which can just use `require('folktale')` and then get the other objects/functions from that root module as properties, but might be a concern for someone who's writing a Browser application. Tree-shaking would help here, but you can't have tree-shaking and first-class modules, and ES6 modules (which would support this) are not finished yet. We might move to ES6 modules when they're finished, and revisit this organisation.

And yeah, one of the reasons to move all of the functionality to a single code base was so this functionality can be reused across the project. So partial functions would return Either or Maybe values, etc.

**Comment by boris-marinov** [comment]
I see. 

Sorry, I got confused since I thought that `core` in your prev example refers to the core of the whole library, and you want to have those as first class functions (obviously I wasn't paying attention). Hence what I said in my response.

Anyway, it's done now. 
And I am ready to pick up something else :)

**Comment by robotlolita** [comment]
No worries, and thanks :)

**Pull request merged**